src/main/java/net/datafaker/providers/base/PhoneNumberGenerator.java

```diff
@@ -0,0 +1,89 @@
+package net.datafaker.providers.base;
+
+import com.google.i18n.phonenumbers.NumberParseException;
+import com.google.i18n.phonenumbers.PhoneNumberUtil;
+import com.google.i18n.phonenumbers.PhoneNumberUtil.PhoneNumberFormat;
+import com.google.i18n.phonenumbers.PhoneNumberUtil.PhoneNumberType;
+import com.google.i18n.phonenumbers.Phonemetadata;
+import com.google.i18n.phonenumbers.Phonenumber;
+import net.datafaker.service.FakeValuesService;
+import net.datafaker.service.FakerContext;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.Objects.requireNonNull;
+
+class PhoneNumberGenerator {
+    private static final PhoneNumberUtil libPhoneNumber = PhoneNumberUtil.getInstance();
+    private static final Map<CacheKey, String> CACHE = new ConcurrentHashMap<>();
+    private static final int MAX_RETRIES = 100;
+
+    private final FakeValuesService fakeValuesService;
+    private final FakerContext context;
+
+    PhoneNumberGenerator(FakeValuesService fakeValuesService, FakerContext context) {
+        this.fakeValuesService = fakeValuesService;
+        this.context = context;
+    }
+
+    String randomPhoneNumber(String countryCodeIso2, PhoneNumberType type, PhoneNumberFormat format) {
+        Phonenumber.PhoneNumber phoneNumber = randomPhoneNumber(countryCodeIso2, type);
+        return libPhoneNumber.format(phoneNumber, format);
+    }
+
+    private Phonenumber.PhoneNumber randomPhoneNumber(String countryCodeIso2, PhoneNumberType type) {
+        String pattern = CACHE.computeIfAbsent(new CacheKey(countryCodeIso2, type),
+            (key) -> phoneNumberPattern(countryCodeIso2, type));
```

Comment:
Is `CACHE` really a constant if it's modified/filled after instantiation?

Reply:
Not sure if I understand your question correctly.
Yes, `CACHE` is `static final` field (of type `Map`). We don't want to assign it any other value.
But we want to store more and more values in this map.

Reply:
If it's are being added to it over time then it's not really a constant 😉 it isn't really "final" (it's not preloaded with content)

Reply:
Well... It's a widely used pattern in Java..
For example, look at the existing class `net.datafaker.annotations.FakeResolver`:

```java
public class FakeResolver<T> {
    private static final Map<Class<?>, FakeResolver<?>> CLASS_2_FAKE_RESOLVER = new CopyOnWriteMap<>(IdentityHashMap::new);

    private static final Map<Class<?>, Schema<Object, ?>> DEFAULT_SCHEMA_CACHE = new CopyOnWriteMap<>(IdentityHashMap::new);
```

Reply:
Alrighty, I don't like it but fine, you're right the code base already uses it.